clients/python/test/e2e/ci/_data_classes.py

```diff
@@ -0,0 +1,172 @@
+            "Either a version or a repo, branch pair must be specified. Not both."
+        )
+        # check at least one is empty
+        if not (
+            is_empty(self.osparc_client_version)
+            or (
+                is_empty(self.osparc_client_repo)
+                and is_empty(self.osparc_client_branch)
+            )
+        ):
+            raise ValueError(msg)
+        # check not both empty
+        if is_empty(self.osparc_client_version) and (
+            is_empty(self.osparc_client_repo) and is_empty(self.osparc_client_branch)
+        ):
+            raise ValueError(msg)
+        if is_empty(self.osparc_client_version):
+            if (
+                is_empty(self.osparc_client_repo)
+                or is_empty(self.osparc_client_branch)
+                or is_empty(self.osparc_client_workflow)
+                or is_empty(self.osparc_client_runid)
+            ):
+                raise ValueError(msg)
+        return self
+
+    @property
+    def version(self) -> Optional[str]:
+        return self.osparc_client_version
+
+    @property
+    def repo(self) -> Optional[str]:
+        return self.osparc_client_repo
+
+    @property
+    def branch(self) -> Optional[str]:
+        return self.osparc_client_branch
+
+    @property
+    def workflow(self) -> Optional[str]:
+        return self.osparc_client_workflow
+
+    @property
+    def runid(self) -> Optional[str]:
+        return self.osparc_client_runid
+
+    @property
+    def compatibility_ref(self) -> str:
+        """Returns the reference for this client in the compatibility table"""
+        if not is_empty(self.version):
+            return "production"
+        else:
+            assert isinstance(self.branch, str)
+            return self.branch
+
+    @property
+    def client_ref(self) -> str:
+        """Returns a short hand reference for this client"""
+        if not is_empty(self.version):
+            assert isinstance(self.version, str)
+            return self.version
+        else:
+            assert isinstance(self.branch, str)
+            return self.branch
+
+
+class PytestConfig(BaseModel):
+    """Holds the pytest configuration
+    N.B. paths are relative to clients/python/test/e2e
+    """
+
+    env: str
+    required_plugins: str
+    addopts: str
+
+
+class Artifacts(BaseModel):
+    artifact_dir: Path
+    result_data_frame: Path
+    log_dir: Path
+
+
+class PytestIniFile(BaseModel):
+    """Model for validating the .ini file"""
+
+    pytest: PytestConfig
+    client: ClientConfig
+    server: ServerConfig
+    artifacts: Artifacts
+
+    @classmethod
+    def read(cls, pth: Path = _PYTEST_INI) -> "PytestIniFile":
+        """Read the pytest.ini file"""
+        if not pth.is_file():
+            raise ValueError(f"pth: {pth} must point to a pytest.ini file")
+        obj = configparser.ConfigParser()
+        obj.read(pth)
+        config_dict: Dict = {s: dict(obj.items(s)) for s in obj.sections()}
```

Comment:
MINOR: `config_dict`: why to name variables with the type if there are already type annotations?

Reply:
I didn't write this. ChatGPT did? I can try to pass on your question to it. 
Joke aside, I will correct